internal/oci/store/store.go

```diff
@@ -158,8 +158,18 @@ func (i *Image) Image(h ociv1.Hash) (ociv1.Image, error) {
 	return oi, errors.Wrap(validate.Image(oi, validate.Fast), errInvalidImage)
 }
 
-// WriteImage writes the supplied image to the store.
+// WriteImage writes the supplied image to the store, fully validating it first.
 func (i *Image) WriteImage(img ociv1.Image) error {
+	// fully validate, not using validate.Fast, the image before writing it to the store.
+	if err := validate.Image(img); err != nil {
+		return errors.Wrap(err, errInvalidImage)
+	}
```

Comment:
I think this is going to make us pull the entire image every time, even if it already exists in the store.

Reply:
See latest change, I moved the validation to the latest stage possible, I had to remove two test cases `WriteLayerError` and `SuccessfulWrite` because otherwise I would have had to mock the whole Image interface and I didn't feel they were actually testing writing. Let me know what you think.

Reply:
I thought about adding validation as an option to be disabled in tests or injecting a validator in the `Image`, but I couldn't find a good API, adding anything to the Image struct didn't feel right, but I didn't want to do a major rewrite of `Image`